---
title: "Database Formatting for DADA2"
author: "Jonathan Lin"
date: "2/14/2019"
output:
  html_document: default
  pdf_document: default
---

This script uses R to modify a curated-curated sequence database (DictDB) into a training fasta file appropriate for assignTaxnomy() in DADA2


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

I will be using a termite-specific curated database called [DictDB](https://www.sciencedirect.com/science/article/pii/S0723202015001162) to classify and assign taxonomy for my termite 16S reads. This database improves taxonomy assignments for termite 16S sequences compared to the standard training sets from SILVA or RDP.

Download DictDB (version 3.0) [here](https://drive.google.com/file/d/0B5NJB7U2TQKNVEZneUhFODZFdmc/view)

Unzip the files and move the 2 files (fasta and taxonomy file) into a new folder named "DictDB" on the desktop, and navigate to that directory.

#### View and inspect the DictDB fasta and taxonomy files:
```{bash}
head -6 DictDB_V3.fasta

head -6 DictDB_V3.taxonomy
```

The fasta file contains the taxa id followed by the sequence after a line. The taxonomy file contains the same taxa id with its associated taxonomy string all on the same line, in the same order as the taxa id's in the fasta file.

Thus, the fasta file should have exactly twice the number of lines as the taxonomy file:
```{bash}
wc -l DictDB_V3.fasta
wc -l DictDB_V3.taxonomy
```
This checks out.

#### To make sure the taxa id strings are matching in both files, add a ">" to the front of every taxa id in the taxonomy file:
```{bash}
sed -e 's/^/>/' DictDB_V3.taxonomy > DictDB_V3_modified.taxonomy

head -6 DictDB_V3_modified.taxonomy
```

With the ">" in front of both taxa id strings, this will allow us to to do string matching. It also preserves the fasta format later on.

#### Load libraries in R, set directory, and read in fasta and taxonomy files as dataframes:
```{r}
library(stringr)
library(dplyr)

setwd("/Users/jonathanylin/Desktop/DictDB")

fasta <- read.table("DictDB_V3.fasta")
taxonomy <- read.table("DictDB_V3_modified.taxonomy")
```

#### View and inspect loaded files:
```{r}
head(fasta)
head(taxonomy)
```

Note that for the dataframe of the fasta file, the taxa id and accompanying sequence are read in the same column, but as separate rows. For the taxonomy file, the taxa id and matching taxonomy string are on the same rows in two different columns.

The assignTaxonomy() command in DADA2 expects a fasta training file in the following format:

```
>Kingdom;Phylum;Class;Family;Genus;Species
sequence
>Kingdom;Phylum;Class;Fmaily;Genus;
sequence
```

We'll need to search through the taxonomy and fasta files and pull out the matching taxa ids with the taxonomy strings.

#### Create a new dataframe with the same number of rows and columns as the taxonomy file:
```{r}
combined <- select(taxonomy, V1)
combined$V1 <- NA
combined$V2 <- NA
```

#### Run the following script:
- Iterates through taxonomy and fasta files; pastes matching taxa string with ">" in new dataframe "combined"
- Pastes matching sequence for each taxa string to second column (same row) in the dataframe "combined"
- Note: It takes about 1 hour to iterate through 55,000+ lines

```{r}
for (i in taxonomy$V1) {
  for (ii in fasta$V1) {
    if (i == ii) {
      combined$V1[which(taxonomy$V1 == i)] = paste(">", taxonomy$V2[which(taxonomy$V1 == i)], sep = "")
      R = which(fasta$V1 == ii) + 1
      combined$V2[which(taxonomy$V1 == i)] = paste(fasta$V1[R], sep ="")
    }
  }
}
```

#### View combined dataframe:
```{r}
head(combined)
```

The taxonomy strings, now each with a ">", should be in a column. The matching sequence for each taxonomy string should also be in the next column.

Now we need to combine these two columns into a fasta-ready format, where the taxonomy string is followed with the sequence:
```{r}
combined_fasta <- do.call(rbind, lapply(seq(nrow(combined)), function(i) t(combined[i, ])))

head(combined_fasta)
```

Now we can export the reformatted dataframe as a fasta file:
```{r}
write.table(combined_fasta, "DictDB_DADA2_FINAL.fasta", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

The file is now ready to use as custom training set for the assignTaxonomy() step in DADA2
